subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileFilesFactory.java

```diff
@@ -131,6 +136,7 @@ private FileVisitResult visitFile(File file, List<IncludeDirectives> visibleIncl
             visibleIncludeDirectives.add(fileDetails.directives);
 
             IncludeFileResolutionResult result = IncludeFileResolutionResult.NoMacroIncludes;
+            boolean hasUnresolvedHeaders = false;
```

Comment:
What is the used of this local variable?

Reply:
We need to distinguish between source files which have unresolved includes and header files which have unresolved includes.
The local variable comes into play when we want to ignore unresolved includes in headers but not in source files. The result of the evaluation is returned and actually used here: https://github.com/gradle/gradle/blob/2433126c24be7f6aee1cd4bad21d9e1eb82a26cc/subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileFilesFactory.java#L101-L102
If we don't ignore unresolved includes in headers, then the variable is unused, as you said, and we use the field instead.

Reply:
But why can't we use the filed in both cases? There is a single result of all this: whether or not we consider some of the files to be unresolved. That should be represented by a single boolean. I don't see why we need a second one if the flag is there.

Reply:
That is the point: There is no single result for `visitFile`: there is a top level result for the source file, which should be written to the field, and there are results of the dependencies of the top level source file, which may or may not be written to the field, depending on the introduced flag. Note the recursive call here: https://github.com/gradle/gradle/blob/2433126c24be7f6aee1cd4bad21d9e1eb82a26cc/subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileFilesFactory.java#L156

Reply:
I removed the local variable and introduced a parameter to `visitFile`. Better now?